enarx-keep-sev-shim/src/hostcall.rs

```diff
@@ -0,0 +1,223 @@
+    /// unsafety in code that relies on it being true.
+    pub unsafe fn from_raw_fd(fd: libc::c_int) -> Self {
+        Self(fd)
+    }
+}
+
+/// Communication with the Host
+pub struct HostCall<'a>(&'a mut Block);
+
+impl<'a> HostCall<'a> {
+    /// Initialize the HostCall machinery with the address of the shared page
+    ///
+    /// # Panics
+    ///
+    /// Panics, if it is initialized more than once.
+    #[inline(always)]
+    pub fn init(address: ShimVirtAddr<*mut Page>) {
+        let prev = unsafe {
+            HOSTCALL_VIRT_ADDR.swap(
+                Address::<u64, *mut Page>::from(address).raw(),
+                Ordering::Release,
+            )
+        };
+        if prev != 0 {
+            panic!("HostCall already initialized");
+        }
+    }
+
+    /// Acquire a HostCall instance to communicate with the host.
+    ///
+    /// Returns None, if it is already in use, which might happen in an interrupt;
+    /// or if there are multiple CPUs.
+    ///
+    /// # Safety
+    ///
+    /// exclusive mutable access is ensured via an Atomic variable
+    #[inline(always)]
+    pub fn try_lock() -> Option<Self> {
+        // Try to take out the virtual address out of the global variable
+        let address = unsafe { HOSTCALL_VIRT_ADDR.swap(0, Ordering::Acquire) };
+        if address == 0 {
+            // Already in use
+            None
+        } else {
+            Some(unsafe { Self(&mut *(address as *mut Block)) })
+        }
+    }
+
+    /// Acquire a HostCall instance to communicate with the host.
+    ///
+    /// # Caution
+    ///
+    /// Do not call this in interrupts, as it can cause deadlocks
+    ///
+    /// # Safety
+    ///
+    /// exclusive mutable access is ensured via an Atomic variable
+    #[inline(always)]
+    pub fn lock() -> Self {
+        loop {
+            // Try to take out the virtual address out of the global variable
+            let address = unsafe { HOSTCALL_VIRT_ADDR.swap(0, Ordering::Acquire) };
+            if address == 0 {
+                // Already in use
+                core::sync::atomic::spin_loop_hint();
+            } else {
+                return unsafe { Self(&mut *(address as *mut Block)) };
+            }
+        }
+    }
+
+    /// Causes a `#VMEXIT` for the host to process the data in the shared memory
+    ///
+    /// Returns the contents of the shared memory reply status, the host might have
+    /// written.
+    ///
+    /// # Safety
+    ///
+    /// The parameters returned can't be trusted.
+    #[inline(always)]
+    unsafe fn hostcall(&mut self) -> Result<[Register<usize>; 2], libc::c_int> {
+        let mut port = Port::<u16>::new(SYSCALL_TRIGGER_PORT);
+        port.write(1 as u16);
+        self.0.msg.rep.into()
+    }
+
+    /// Write `bytes` to a host file descriptor `fd`
+    ///
+    /// Write at most `Block::buf_capacity()` bytes.
+    /// Handle it like write(2) and call it in a loop until all bytes are written.
```

Comment:
This comment is out of date.

Reply:
No, I just changed it :) Should reflect the method.